### DISCUSSION TITLE
请问ColossalAI支持offload的操作吗

### DISCUSSION THREAD

**Discussion by xxw11** [hpcaitech/ColossalAI#207] [Community | Q&A] [answered]
请问显存不足的情况可以怎么处理？ColossalAI支持如 cpu offload的之类的操作吗？

**Comment by FrankLeeeee** [comment, marked as answer]
Hi，we support offload. You can read the documentation at [https://colossalai.org/tutorials/features/zero_redundancy_and_zero_offload](https://colossalai.org/tutorials/features/zero_redundancy_and_zero_offload)

**Reply by xxw11** [reply]
敬爱的开发者，十分感谢你的回答，但我还有一些困惑在文档中似乎没有找到：
1.ColossalAI 支持像 gradient/activation checkpointing 这样减少显存的操作吗？
2.ColossalAI中cpu/nvme offload 可以与 1d/2d/3d 并行策略一同使用吗？（因为给出的example中好像都是单独的）

**Reply by FrankLeeeee** [reply]
Hi，
1. 支持，函数可以在[api文档](https://colossalai.readthedocs.io/en/latest/colossalai/colossalai.utils.html)里找到。谢谢反馈，我们会在tutorial里加入一个checkpointing的例子。
2. 可以一起使用

**Reply by xxw11** [reply]
十分感谢你的耐心回复，刚才阅读ColossalAI的paper，对论文最后Memory Efficiency举的例子，用4个p100（16GB）训练3B大小的模型不需要offload操作而感到十分震惊，因为我用4个rtx 3090 (24GB) 训练5B大小左右的GPT2 严重依赖于offload，能否提供关于这个例子的更多技术细节，再次感谢！
![image](https://user-images.githubusercontent.com/55141908/153372939-dd3d9231-843f-46ea-94a6-32064a820eff.png)

**Reply by xxw11** [reply]
4.72B的gpt2

**Reply by FrankLeeeee** [reply]
你好，我们目前在写tensor parallel部分的tutorial，但是由于并行种类较多，所以仍在进行中。我们写完之后会更新到tutorial文档的features目录下。我们的[例子库](https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt)里提供了GPT的代码，你可以试一下用1D/2D模式跑一下看看。paper里的应该是基于ViT的结果，所以会有出入。

**Reply by xxw11** [reply]
请问ColossalAI有提供像megatron对使用xxx.idx xxx.bin这种已处理后的数据集的方法吗？查看tutorials和docs后，似乎没有找到

**Reply by FrankLeeeee** [reply]
Hi，抱歉回复晚了，目前并没有提供这样的功能，不过应该替换一下dataset类就可以？

**Reply by FrankLeeeee** [reply]
https://github.com/hpcaitech/ColossalAI-Examples/tree/main/language/gpt
这个例子里的GPT也用了megatron的数据处理方法，应该能和你的兼容

**Reply by xxw11** [reply]
十分感谢 ，在源代码基础上修改了一些，已经解决了，十分感谢

**Comment by upwindflys** [comment]
您好，上面得页面找不到了，请问现在哪个优化器支持offloading呢
